Add limited support for electrum segwit seeds

BamaHodl opened the pull request:
## Description

Add support for electrum segwit seeds (Issue #438  ) so that users with electrum seeds will not need to move utxos to a BIP-39 seed to use SeedSigner as their signing device.

It requires one new option to enable Electrum seed support in Advanced settings (disabled by default), which will allow the user the option to enter Electrum seed phrases on screens that allow the user to enter BIP-39 seed phrases.  Per the PR comments, SeedQR has been disabled for Electrum seeds.

This pull request is categorized as a:

- [x] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [x] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [x] Yes
- [ ] N/A

I have tested this PR on the following platforms/os:

- [ ] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [x] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [x] Other - seedsigner-emulator

jdlcdl commented:
I've taken a first look at this pr as well as a single test:

* Can recover a 12 word electrum seed, was offered to switch to "electrum", chose "custom extension" and recovered the correct bip32 wallet with my 13th word.
* was able to backup via seed words.
* was able to backup via 25x25 standard SeedQR (because compact couldn't possibly work w/o checksum bits).

much more to check, but this is interesting for many I suspect.  ty @BamaHodl

jdlcdl commented on src/seedsigner/helpers/embit_utils.py:
```diff
@@ -30,6 +30,8 @@ def get_standard_derivation_path(network: str = SettingsConstants.MAINNET, walle
         raise Exception("Unexpected network")
 
     if wallet_type == SettingsConstants.SINGLE_SIG:
+        if is_electrum:
```
Wondering if this (and the multisig one below) should be nested under the native-segwit branch just below?

BamaHodl replied:
Yes, this may be the best route considering there may be a way to enable nested segwit (currently possible, just need to figure out how) and taproot (not currently supported in electrum) for electrum seeds.

Will make this suggested change with some other mods.

BamaHodl replied:
this change has been made

jdlcdl commented:
As of commit #89f6cec,

I have reviewed, and tested, nothing newly broken.

I still confirm that Address Verification appears to search standard native-segwit bip84 path instead of m/0', but I understand that for electrum seeds, not all functionality is expected to work.

BamaHodl commented:
> For Address Verification, it appears to search standard native-segwit bip84 path instead of m/0', but I understand that for electrum seeds, not all functionality is expected to work.

My mistake, I was unaware it was not working for electrum seeds.  The way the derivation path is set for address verification is to take a default derivation path for bip-39 seeds rather than looking it up based on the seed itself so I missed that one.  A simple fix was made to simply check if the seed is electrum before doing the search, and override the derivation path if so.  Address verification works with this latest change.

jdlcdl commented:
> Address verification works with this latest change.

I confirm.

Very nice BamaHodl!

jdlcdl commented:
Without testing and just with a glance of [289b382](https://github.com/SeedSigner/seedsigner/pull/513/commits/289b382720bfce4bcdf2374c691657bba7209b60)

I wonder if that commit, all by itself, might better be a refactoring pull-request separate from this one.

I just say that because it's sort of off topic from "electrum support", even if it does reduce the need for "is_electrum", and smaller pull-requests are easier to review.  I'd hate to see some progress not go forward because it were waiting for other progress to be reviewed and accepted.

Up to you.   Thanks again for nice improvements.

BamaHodl commented:
I see your reasoning from the procedure aspect--should probably be a PR on its own.  Was just trying to clean it up a bit to avoid needing to do electrum-specific stuff there.
I've backed out that commit

jamesturnernz commented:
Can confirm that this would be a useful comparability feature with a popular open source wallet.  Being able to bring people into SS from Electurm would be useful.  Thank you

kdmukai commented:
@BamaHodl, can you find a sensible spot in the docs to add info about Electrum seed import? Especially clarification about which Electrum format is supported (I had no idea they updated to a version with the bip39 wordlist; should say that we don't support their older wordlist). Also note any other limitations/caveats when in this mode.

kdmukai commented:
I'm very uncomfortable with the idea of facilitating SeedQR creation for an Electrum seed. I can confirm that it does work to scan back in. But it somewhat muddies the water of what a SeedQR encodes and would likely be wholly unsupported by other software/hardware that has adopted the SeedQR standard.

It seems like EITHER we:
* push hard for Electrum seeds to be explicitly part of the SeedQR standard and advocate that all other implementations do the same (not my preference)
* OR we, as the reference SeedQR implementation, keep the standard simple and focused.

My main interest in supporting this PR is to give users with Electrum seeds the ability to use their seed in SeedSigner, BUT if it's a clunkier, less convenient experience for them, so be it.

SeedSigner w/Electrum seed makes the most sense for rare cases of emergency signing / bailing you out of trouble. And to a lesser extent, external verification (e.g. address explorer). But Electrum seed usage in general should be discouraged. And maybe even more functionality should be disabled (e.g. a bip85 child seed derived from an Electrum seed parent will still yield a bip39 seed... or we can just say, "No, not even gonna go there").

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -427,6 +427,21 @@ def __post_init__(self):
         )
         self.components.append(self.fingerprint_icontl)
 
```
Our opinionated / slightly non-standard formatting: 3 blank lines between classes or other elements at the top nesting level, 2 blank lines between functions or other elements at nested levels.

We're not fully consistent with that everywhere, but mostly so.

BamaHodl replied:
I will make the spacing consistent--was just an oversight

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -427,6 +427,21 @@ def __post_init__(self):
+    def __post_init__(self):
+        self.show_back_button = False
+        self.title = "Switch to Electrum?"
+        self.is_bottom_list: bool = True
```
"Switch" isn't quite the right verb here. 

Would "Electrum Seed Detected" fit in the title? If not, the title can just say "Electrum Seed" since the body text will have more info.

BamaHodl replied:
Same response as below--agree on removing switch language but it is in a tiny doubt if the user meant to enter BIP-39 or electrum.  Alternatively we could check if it meets both checksums and only have a confirmation screen in that edge case, and go with the Electrum Seed declaration and warning if it is a valid electrum seed but not a valid BIP-39 checksum.  Does that sound good?  Confirmation screen if it meets both checks, warning screen if it's electrum-only

BamaHodl replied:
Yes, It's making more sense to me too, especially after learning the collisions are more often than I had initially thought.  And in this way it keeps the BIP-39 flow completely unaffected by default even when they happen to collide with a valid Electrum seed (apparently 9/2048 chance which is much higher than I had calculated).
This also removes the need for a new screen altogether possibly, at least during seed entry

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -297,9 +300,26 @@ def __init__(self):
+                self.seed.switch_to_electrum()
+                # recalculate fingerprint
+                self.fingerprint = self.seed.get_fingerprint(network=self.settings.get_value(SettingsConstants.SETTING__NETWORK))
+            elif not self.seed.seed_bytes: 
```
Missing an `elif button_data[selected_menu_num] == self.CANCEL` branch.

BamaHodl replied:
It falls through and behaves correctly (i.e. tries to treat it as BIP-39), but agree it can be more clear at least with a elif with a comment there

kdmukai replied:
If a user select "Cancel", they're explicitly saying, "Oh, never mind, I don't want to load this Electrum seed." So I think the better behavior would be to discard the seed and just return us to Home.

BamaHodl replied:
I see what you mean now.  I had intended the flow to be -- if it's valid electrum seed, ask the user if they intended that and if not, then fall through and try to treat it as BIP-39.  So in this new case it should have 3 buttons then--treat as Electrum, treat as Bip-39, and cancel (or just remove cancel altogether).

If the preferred way is to do the 2-step (which I'm leaning more to myself), it could be that the user puts the SS in "Electrum mode" in advanced settings, and then we don't even need an extra screen--only support electrum seeds while in electrum mode and only support bip-39 otherwise, e.g. behavior doesn't change at all for bip-39 seeds in default mode even if their seed collides with a valid electrum seed.  Thoughts?

kdmukai replied:
> only support electrum seeds while in electrum mode and only support bip-39 otherwise

That's too restrictive for me. In my mind, SeedSigner has a core bip39 foundation. Electrum seed support may be added on the side, but it should never override/exclude the main bip39 use cases. That's why I like the explicit "Load Electrum Seed" option; it provides a single, specific path for this one side option.

And if you go through the normal bip39 mechanisms and load a mnemonic that could also be a valid Electrum seed, it doesn't matter, we should never even check the Electrum side in the normal flows. Doesn't exist.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -427,6 +427,21 @@ def __post_init__(self):
+        super().__post_init__()
+
+        self.components.append(TextArea(
+            text="It appears this is an Electrum-style seed, switch it to be so? Some functions are not supported for Electrum seeds",
```
I wouldn't say "It appears". It IS an Electrum seed, as far as we can tell. "switch" still bugs me.

This may be too extreme, but I'm slightly more inclined to a more drastic presentation:

<img width="332" alt="Screenshot 2024-03-03 at 5 17 44 PM" src="https://github.com/SeedSigner/seedsigner/assets/934746/96441074-0256-40bd-8233-8e4d674c668e">

BamaHodl replied:
Agree we can make the warning more extreme and remove the 'switch' wording.  But there is a tiny possibility it is in fact a BIP-39 seed (1 in 2^24), which is why I put a confirmation screen.  more here:
https://github.com/SeedSigner/seedsigner/pull/513#issuecomment-1976732604

kdmukai commented:
I haven't fully thought this through, but I feel like we should be able to isolate all of the Electrum-specific code in seed.py by subclassing `Seed`:

```python
class Seed:
    ...

class ElectrumSeed(Seed):
    ...
```

That would then shift the "is it bip39 or is it Electrum" logic to https://github.com/BamaHodl/seedsigner/blob/dev/src/seedsigner/models/seed_storage.py#L90-L91 where if the `Seed` instantiation fails, we can try `ElectrumSeed` instead.

Also here if we support Electrum SeedQRs:
https://github.com/BamaHodl/seedsigner/blob/dev/src/seedsigner/views/scan_views.py#L75-L77

Perhaps a higher-level util could further abstract that away (e.g. a `SeedParser` in seed.py that returns either a `Seed` or an `ElectrumSeed`).

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1650,6 +1678,9 @@ def __init__(self, seed_num: int = None):
         self.address = self.controller.unverified_address["address"]
+        # override derivation path if it's electrum as bip-39 path was set by default
+        if self.seed and self.seed.is_electrum:
+            self.controller.unverified_address["derivation_path"] = 'm/0h'
```
The hard-coding here and the limited Electrum support in `embit_util.get_standard_derivation_path()` seems pretty awkward, but I don't have an improvement suggestion yet.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -612,7 +633,9 @@ def __init__(self, seed_num: int, sig_type: str):
         args = {"seed_num": self.seed_num, "sig_type": self.sig_type}
-        if len(self.settings.get_value(SettingsConstants.SETTING__SCRIPT_TYPES)) == 1:
+        seed = self.controller.storage.seeds[self.seed_num]
+        script_types = [SettingsConstants.NATIVE_SEGWIT] if seed.is_electrum else self.settings.get_value(SettingsConstants.SETTING__SCRIPT_TYPES)
```
This confused me a bit as a user. What's the rationale for not supporting other single sig script types w/Electrum seeds? And if it has to stay limited to Native Segwit, I think a notification screen might be needed (e.g. "Note: Only native segwit is supported for Electrum seeds" w/a "Next" button).

BamaHodl replied:
In electrum wallet the versioning indicates the script type.  Ascii "100" version means native segwit and only native segwit.  I may be able to add legacy seeds eventually as well, but I've only worked on native segwit version so far, so allowing that option wouldn't be useful until supporting legacy seed types.  There is no electrum seed version for taproot yet either.

kdmukai replied:
I still think we need to tell the user that right now they're only going to get native segwit addrs. Then later if/when more options are available, it's the normal button list. Consider the user who's trying to get legacy types but doesn't realize the limitation. It'd be pure frustration that the UI just skips straight to native segwit.

BamaHodl commented:
>  1/16 of them is also a valid BIP39 seed

I believe it's more like 1 out of 16 million?  Basically, if the hash of the seed starts with "100" in ascii, not binary (e.g. 3 full bytes must match, 1 in 2^24 probability)

Also, the reverse is not possible--electrum-generated seeds will never have a BIP-39 checksum because there is a check to discard those when electrum wallet generates seeds to avoid that confusion.  But you are correct that 1 in 2^24 BIP-39 seeds will be a valid electrum seed--that's why I made it require user confirmation rather than just assuming they meant to enter an electrum seed.  It's a very small chance, but we don't want to skip any edge cases by making assumptions for sure

3rdIteration commented:
> > 1/16 of them is also a valid BIP39 seed
> 
> I believe it's more like 1 out of 16 million? Basically, if the hash of the seed starts with "100" in ascii, not binary (e.g. 3 full bytes must match, 1 in 2^24 probability)
> 
> Also, the reverse is not possible--electrum-generated seeds will never have a valid BIP-39 checksum because there is a check to discard those when electrum wallet generates seeds to avoid that confusion. But you are correct that 1 in 2^24 BIP-39 seeds will be a valid electrum seed--that's why I made it require user confirmation rather than just assuming they meant to enter an electrum seed. It's a very small chance, but we don't want to skip any edge cases by making assumptions for sure

Electrum only added this check in mid 2020. (And other wallets which generate Electrum compatible seeds may not check it at all)

BamaHodl commented:
> > > 1/16 of them is also a valid BIP39 seed
> > 
> > 
> > I believe it's more like 1 out of 16 million? Basically, if the hash of the seed starts with "100" in ascii, not binary (e.g. 3 full bytes must match, 1 in 2^24 probability)
> > Also, the reverse is not possible--electrum-generated seeds will never have a valid BIP-39 checksum because there is a check to discard those when electrum wallet generates seeds to avoid that confusion. But you are correct that 1 in 2^24 BIP-39 seeds will be a valid electrum seed--that's why I made it require user confirmation rather than just assuming they meant to enter an electrum seed. It's a -very small chance, but we don't want to skip any edge cases by making assumptions for sure
> 
> Electrum only added this check in mid 2021. (And other wallets which generate Electrum compatible seeds may not check it at all) Basically any segwit compatible electrum seed generated before then has a 1/16 chance of being a valid BIP39 seed too.
> 
> See here: [spesmilo/electrum#6001](https://github.com/spesmilo/electrum/pull/6001)

Thanks, was unaware of the history and assumed the check was there was since integrating segwit.  There are many more electrum seeds that are also valid BIP-39 seeds than I had guessed, then, since the check was not always there.  There are still many fewer BIP-39 seeds that are valid electrum seeds, however even so it is also much higher than I had estimated.  According to that discussion it's (0.439%) and not 1/2^24

BamaHodl commented:
> I haven't fully thought this through, but I feel like we should be able to isolate all of the Electrum-specific code in seed.py by subclassing `Seed`:
> 
> ```python
> class Seed:
>     ...
> 
> class ElectrumSeed(Seed):
>     ...
> ```
> 
> That would then shift the "is it bip39 or is it Electrum" logic to https://github.com/BamaHodl/seedsigner/blob/dev/src/seedsigner/models/seed_storage.py#L90-L91 where if the `Seed` instantiation fails, we can try `ElectrumSeed` instead.
> 
> Also here if we support Electrum SeedQRs: https://github.com/BamaHodl/seedsigner/blob/dev/src/seedsigner/views/scan_views.py#L75-L77
> 
> Perhaps a higher-level util could further abstract that away (e.g. a `SeedParser` in seed.py that returns either a `Seed` or an `ElectrumSeed`).

I had also considered this, but decided against it, trying to recall why..  I think it was just because I wanted to minimize changes and avoid doing anything that would be seen as changing the architecture of the code (e.g. by moving functionality into the seed class and Electrum seed sub-class).  I could take another look at this though if it's an acceptable way to change things and see if it simplifies the code a lot.  Will get back on this.

kdmukai commented:
> I had also considered this, but decided against it, trying to recall why.. I think it was just because I wanted to minimize changes and avoid doing anything that would be seen as changing the architecture of the code (e.g. by moving functionality into the seed class and Electrum seed sub-class).

My personal preference is to keep any Electrum-specific code off to the side as much as possible. The new code would ideally be a parallel drop-in rather than seeing `if seed.is_electrum:` conditionals everywhere. But within reasonable limits of DRY principles, of course.

Thus this suggestion of a `ElectrumSeed` child class where there'd be almost zero duplication and obviously meets my goal of keeping those conditionals out of the `Seed` class.

BamaHodl commented:
For clarity, my understanding of feedback has led to the following plan:

- Remove SeedQR support for Electrum seeds, largely because only SS would support it
    - An Electrum SeedQR could cause confusion, especially in an unplanned inheritance (heirs scan QR and just get invalid checksum on any other signing device)
    - This also simplifies this PR by eliminating some of the needed code changes to support QR load/backup for Electrum seeds
    - This also eliminates the need to ever ask the user if they intended to enter BIP-39 or Electrum in the event a seed scanned from the main "Scan" menu is valid for both
- A new Advanced Setting option to enable Electrum seed support (necessary?), disabled by default
- A new option in the "Load A Seed" menu to "Enter Electrum Seed" (when Electrum seed support option is enabled)
    - If user enters an unsupported Electrum seed version (i.e. anything other than SegWit), display a NotYetImplemented screen with details
    - Will attempt to add legacy ("Standard", as it's officially named) Electrum seed version support as part of these additional changes if it can be easily done 
- Create separate ElectrumSeed (parent: Seed) class and to the extent possible use polymorphism to handle differences between Electrum and BIP-39 seed behavior
    - This may require moving some functionality in other classes into the Seed and ElectrumSeed classes
- If user attempts to "Export as QR" on an ElectrumSeed, display a NotYetImplemented screen with a note like "SeedQR for Electrum seeds is deliberately disabled, see docs"
- Update the docs to elaborate on how to use SS for Electrum seeds what features are and are not supported for Electrum seeds and why
- Update Electrum seed tests to reflect these changes and add tests where appropriate

Please let me know if there is additional feedback on this plan or if I have misunderstood anything.

kdmukai commented:
@BamaHodl excellent summary, great job collecting it all together! Also thank you for being so open to all the suggestions/changes/etc.

> * A new Advanced Setting option to enable Electrum seed support (necessary?), disabled by default

Yes, I think it's necessary. We've tried to keep the UI as simple as possible. I think every additional option creates exponentially more doubt and confusion for a new user (e.g. the jump from 2 options on a screen to 3). The beauty of our Settings is that we can bury TONS of advanced functionality in there so that the majority of users never see the feature in the UI. It's much better to answer "Can it do X?" ("Yes, just go to Settings") vs "I'm confused, do I select X, Y, or Z?"

---

> Will attempt to add legacy ("Standard", as it's officially named) Electrum seed version support as part of these additional changes if it can be easily done

Based on @3rdIteration's comments, are there effectively three(ish) Electrum types?

1. Original w/non-bip39 wordlist
2. w/bip39 wordlist but w/possible inadvertent bip39-valid checksum
3. w/bip39 wordlist, bip39-valid checksum not possible(?)

Though for our purposes, 2.) and 3.) are effectively the same (w/the updated flow forcing them to explicitly tell us they're trying to enter an Electrum seed, we'll never even notice if it happened to also be bip39 compliant), right?

---

> Create separate ElectrumSeed (parent: Seed) class and to the extent possible use polymorphism to handle differences between Electrum and BIP-39 seed behavior

I think this should be a pretty clean child class that only need to override a handful of simple methods.

Also with the updated flow, I don't think we'll need the `SeedParser` concept I mentioned above. The updated flows always know which Seed type to expect, so they can just directly instantiate the appropriate one.

---

More food for thought: if possible, see if creating a child class of any existing `View`s makes sense to further isolate the Electrum-specific code. I'd love it if `SeedExportXpubDetailsView` could get a child class `ElectrumSeedExportXpubDetailsView` to keep all the Electrum mods out of that already more complicated parent View. the parent could, for example, be initialized with default `embit_util` functions to call. But the child could override those to Electrum version equivalents (it would be a win-win to pull the Electrum logic out of those embit_utils functions anyway. The Electrum logic in them is pretty trivial, I think?).

```python
class SeedExportXpubDetailsView(View):
    # I'm not 100% sure about the Callable type hint details...
    get_derivation_path: Callable[..., str] = embit_utils.get_standard_derivation_path

    def run(self):
        derivation_path = self.get_derivation_path(...)
        ...

class ElectrumSeedExportXpubDetailsView(View):
    get_derivation_path: Callable[...,str] = embit_utils.get_electrum_derivation_path

    # Maybe that's it? The entire child class is just 1-2 function definitions...?
```

With that minimal rewiring of which version of the function gets called, maybe no other overrides are needed AND no `if seed.is_electrum` conditionals? I haven't fully tried this out so I could be totally wrong.

---

Another approach could work for `SeedAddPassphraseView`. If the title was an optional arg on `__init__`, the calling View could just set "Custom Extension" so that `SeedAddPassphraseView` would never care which type of Seed it was working with.

Again, not totally sure it's worth it just to avoid one `if seed.is_electrum` test, but the basic point is to look for every opportunity to streamline those away, especially when they intrude into common / important areas of the code.

---

> If user attempts to "Export as QR" on an ElectrumSeed, display a NotYetImplemented screen with a note like "SeedQR for Electrum seeds is deliberately disabled, see docs"

I would just filter the Seed Options list and not even display that button.

I also think other optional features like bip85 should not be displayed. Maybe someone will make a strong case for enabling the combination (Electrum seed creating bip85 children) in a later release, but for now I'd favor a more conservative / restrictive approach.

---

Speaking of that, there wasn't a mention of what to do with the warning screen.

I think the flow should probably be:
* Home -> Seeds -> Enter/Load Electrum Mnemonic (or whatever matches the language for the existing parallel options)
* --> Show warning screen: some features disabled <--
* Enter Electrum mnemonic...

If there's any strong pushback on that warning, perhaps the Settings option for Electrum support could be:
* Disabled (default)
* Enabled w/warning
* Enabled no warning

pointbiz commented:
It would be beneficial to have the Electrum seed specification linked here.
I found this:
https://electrum.readthedocs.io/en/latest/seedphrase.html

Is there another document?

pointbiz commented:
There's another non-standard seed scheme by Lightning Labs LND open source project. 
See here
https://github.com/lightningnetwork/lnd/blob/master/aezeed/README.md

It has versioning and wallet birthdays. 

Electrum seeds are missing birthdays and checksums.

When looking through advanced settings on my 0.7.0 SS I don't see any settings or options like this proposed support for non-standard seeds. 

I'm wondering what derivation paths Electrum seeds use and if this can create foot guns for users who use BIP43/44/49/84/86 derivation paths with Electrum seeds that are supposed to capture that information in the seed version prefix.

Thomas V. made a brilliant invention with mnemonic seeds. Upon peer review versions and derivation paths were not baked in for valid flexibility reasons and a word list was standardized and a checksum was added.

Should SS be like Electrum? A break-glass in case of emergency tool. Or should it be a go to, happy path, best practices tool?

pointbiz commented:
Versioning the seed can lead to the version being ignored. Like if you version an Electrum seed for single-sig standard and then reuse the same seed for multisig. 

Ethereum made this mistake where the public key has a network version and then people reuse the Ethereum address on Rootstock but keep the Ethereum mainnet version number, essentially ignoring the version number.

BamaHodl commented:
Thank you for your questions @pointbiz :

> It would be beneficial to have the Electrum seed specification linked here.
> I found this:
> https://electrum.readthedocs.io/en/latest/seedphrase.html
> 
> Is there another document?

That is a good resource and I also found this excerpt helpful on the history of electrum mnemonics and how they evolved across versions of the software: https://bitcoinelectrum.com/frequently-asked-questions/#why-wont-electrum-accept-my-seed-when-i-attempt-to-restore-my-wallet-from-it
"Old" mnemonics in Electrum software prior to version 2.0 (March 02, 2015) are 12 words from a different wordlist from BIP-39.
"Standard" mnemonics are from the BIP-39 wordlist in Electrum version 2.0+, and until version 2.7.0 (Oct 02, 2016) they were 13 words, and 12 words thereafter.
"Segwit" mnemonics are available in version 3.0+ (Nov 01, 2017) and are 12 words from the BIP-39 wordlist and the address derivation is compatible with BIP-32.  These seeds are what this PR is focused on as they are most compatible with existing SS functionality.  I am still investigating the possibility to add support for the "Standard" non-segwit seed type.

> I'm wondering what derivation paths Electrum seeds use and if this can create foot guns for users who use BIP43/44/49/84/86 derivation paths with Electrum seeds that are supposed to capture that information in the seed version prefix.

Electrum wallet software does not allow selecting derivations for its own mnemonics--they are restricted to "m/0h" as the root node in the wallet software, so that is what is replicated in this PR for the segwit seed version

BamaHodl commented:
@kdmukai I have integrated the changes we discussed as follows:

- Added an ElectrumSeed class that derives from Seed and handles all of the differing functionality in those classes instead of needing is_electrum checks everywhere
- Changed the EncodeQR class to accept a Seed class instance instead of mnemonic and passphrase.  This was necessary to remove the need for EncodeQR classes to re-create the Seed class instance in them and thus eliminated the need to pass is_electrum everywhere
- Electrum seed support is disabled by default, and can be enabled in the Advanced Settings
- When Electrum seed support is enabled, the user is given an option to "Enter Electrum seed" wherever they are given the option to enter BIP-39 seeds.  This means that there will never be confusion as to which seed type the user intended to enter and thus eliminated the need for a new "Switch to Electrum" screen.
- SeedQR support for Electrum seeds is disabled--it is not listed as an option for Electrum seed backup.
- BIP-85 child seeds are disabled for Electrum seeds--it is not listed as an option for Electrum seeds.
- A warning screen is displayed before the user enters an Electrum seed notifying them that some functionality is not supported for Electrum seeds
- The docs have been updated to note that Electrum Segwit seeds are supported, and a new doc page for electrum seeds elaborates further on how to enter an Electrum seed in SS and its limitations

berlinxray commented:
I tested this great addition to Seedsigner today. 
I managed to sign a transaction. 
Unfortunately, Electrum, Sparrow and Bluewallet could't parse the animated QR-Code to import the transaction. 
Looking forward to this valuable addition to Seedsigner!

berlinxray commented:
Finally, I managed to sign an transaction with Sparrow wallet, Seedsigner and an Electrum seed :-) 
Sending the signed transaction via QR-Code to Sparrow was successful, broadcasting too...

My previous problems on May 21 may have been caused by a large transaction size and Electrum having trouble reading animated QR-Codes...

BamaHodl commented:
> Finally, I managed to sign an transaction with Sparrow wallet, Seedsigner and an Electrum seed :-) Sending the signed transaction via QR-Code to Sparrow was successful, broadcasting too...
> 
> My previous problems may have been caused by a large transaction size and Electrum having trouble reading animated QR-Codes... I forgot Electum calls seed passwords a "seed extension". This change does add an option for a custom extension. I apologise for my ignorance.

Thanks for the update, and apologies for not seeing your prior comment!  Glad you were able to both get the tx signing and custom extension features working using Sparrow as your coordinator.  It should be noted that the official Electrum wallet software is not yet a supported coordinator for the SeedSigner, so you are correct that loading and signing transactions with Electrum software will not work currently, in particular because as you mentioned animated QR formats are not yet supported.  For my use what I do is load my Electrum seed into the SeedSigner using this new feature, export the xpub into a supported coordinator software (Sparrow is my favorite), and then everything works as expected as any other seed would.

I am taking another look at this PR and will get it merge-ready again at least from a testing and merge conflict standpoint so that we can hopefully get this in the main repo and a future production release if all goes well.

3rdIteration commented:
@BamaHodl I did some work on adding full legacy support that I think I will bring in another PR. (As there didn't seem to be any objections)

Basically once this is in there, the rest of the legacy Electrum stuff will pretty much "just work" with minimal extra work.

BamaHodl commented:
> @BamaHodl I did some work on adding full legacy support that I think I will bring in another PR. (As there didn't seem to be any objections)

That's awesome!  Just curious what exactly you mean when you refer to legacy, do you mainly mean supporting p2pkh script type?  That was my main reason for not attempting any of the other electrum seed types other than native segwit, namely p2pkh not currently implemented in seedsigner.  If we can get p2pkh support then adding the older electrum seed types should be relatively straight-forward.  These are the electrum seed types I found in my research:

| Name  | Wordlist | Num Words | Script Type | Electrum version |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| Native SegWit | BIP-39 | 12 | p2wpkh | >=3.0 |
| Standard | BIP-39 | 12 | p2pkh or multi-sig p2sh | >=2.7 and <3.0 |
| Standard | BIP-39 | 13 | p2pkh or multi-sig p2sh| >=2.0 and <2.7 |
| "Old" | Legacy | 12 | p2pkh | <2.0 |

jdlcdl commented on docs/electrum.md:
```diff
@@ -0,0 +1,14 @@
+	- Not applicable for Electrum seed types
+- SeedQR backups
+	- Since Electrum seeds are not supported by other SeedQR implementations, it would be dangerous to use SeedQR as a backup tool for Electrum seeds and is thus disabled
+- Custom derivations
```
I can understand not doing this for bip85 because the user might NOT be expecting a bip39 mnemonic.  But wondering why no custom derivation for electrum seeds.

3rdIteration replied:
Electrum doesn't use standard derivation paths, or let users edit them, they are basically hardcoded based on the seed type.

newtonick commented:
@BamaHodl we are planning a pre release soon. Hopefully before the end of July. Would you be able to resolve the merge conflict on this PR (It looks like just a whitespace issue) before I do code review and testing? Thanks!

BamaHodl commented:
> @BamaHodl we are planning a pre release soon. Hopefully before the end of July. Would you be able to resolve the merge conflict on this PR (It looks like just a whitespace issue) before I do code review and testing? Thanks!

Fixed

newtonick commented:
A lesser known feature of SeedSigner is that it can actually scan and [decode](https://github.com/SeedSigner/seedsigner/blob/c94725e7fa6f4f9ef6c3c8fd33e11e454d9a5900/src/seedsigner/models/decode_qr.py#L740) a single frame PSBT QR (Base43) from Electrum. This feature has never been advertised or really supported, but I remember working on it. It wasn't really practical since there was not Electrum supported way to get the signed PSBT back into Electrum.

<img width="926" alt="image" src="https://github.com/SeedSigner/seedsigner/assets/127377/a00bfd63-6cd9-4f3a-8011-7381e9317f62">

newtonick commented:
Overall I'm good with this PR minus a few changes. I really hope @kdmukai finds time to update his review of this PR. I have a feeling there might be some thing he'd want to see updated since all the changes @BamaHodl has made since the beginning of March. I appreciate this great contribution @BamaHodl!

newtonick reviewed (changes requested):
Please add `import` for `WarningScreen` to tools_view.py. Without this dependency an error is thrown in Address Explorer for Electrum seeds.

3rdIteration commented:
I'm happy to add the code to support legacy Electrum addresses in a separate PR (That is trivial once this and the other Legacy address type PR are merged)

3rdIteration commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -458,6 +468,13 @@ class SettingsDefinition:
+        SettingsEntry(category=SettingsConstants.CATEGORY__FEATURES,
+                      attr_name=SettingsConstants.SETTING__ELECTRUM_SEEDS,
+                      abbreviated_name="Electrum",
+                      display_name="Electrum seed support",
```
Electrum never really supported P2SH Segwit seed generation, so it can just be `Electrum Segwit P2WPKH`

BamaHodl replied:
Changed it to indicate it's Native Segwit seeds only.  I put that bit after "Electrum seed support", due to the limitation of how much text displays before selecting the option.  Once the user clicks Electrum seed support they can then see the full description that indicates it's for Native Segwit only.

BamaHodl commented:
> I'm happy to add the code to support legacy Electrum addresses in a separate PR (That is trivial once this and the other Legacy address type PR are merged)

I'm also going to attempt integrating the old Standard seed types that use P2PHK.  Might be able to accomplish full historical electrum seed support now with your changes.  The difficulty with supporting them all will be the UX.  Will have to give the user options to use 12 word seed phrases, 13 word seed phrases (with a warning this is NOT the custom extension), and the option to load from the old, pre-2.0 wordlist (i.e. not BIP-39 wordlist) to call it full coverage.

3rdIteration commented:
> > I'm happy to add the code to support legacy Electrum addresses in a separate PR (That is trivial once this and the other Legacy address type PR are merged)
> 
> I'm also going to attempt integrating the old Standard seed types that use P2PKH. Might be able to accomplish full historical electrum seed support now with your changes. The difficulty with supporting them all will be the UX. Will have to give the user options to use 12 word seed phrases, 13 word seed phrases (with a warning this is NOT the custom extension), and the option to load from the old, pre-2.0 wordlist (i.e. not BIP-39 wordlist) to call it full coverage.

Yea the *really* old seeds are certainly a bit of an issue UX wise. It could be that a separate PR is required which allows users to enable/disable different seed types in a similar manner to how script types are enabled/disabled. (Which would also make it easier to add other things like SLIP39 or other future seed standards which may emerge)

BamaHodl commented on docs/electrum.md:
```diff
@@ -0,0 +1,14 @@
+# SeedSigner Electrum seed phrase support
+
+SeedSigner supports loading of Electrum's Segwit seed phrases.  This is considered an Advanced feature that is disabled by default.  
```
Good idea, done.

BamaHodl commented on src/seedsigner/views/tools_views.py:
```diff
@@ -492,6 +495,18 @@ def run(self):
 
+        elif button_data[selected_menu_num] == self.TYPE_ELECTRUM:
+            self.run_screen(
+                    WarningScreen,
```
Good catch, done.

BamaHodl commented:
> I'm happy to add the code to support legacy Electrum addresses in a separate PR (That is trivial once this and the other Legacy address type PR are merged)

playing around with the older electrum p2pkh seeds here with the addition of your legacy script support PR:  https://github.com/BamaHodl/seedsigner/tree/LegacyElectrum 

I haven't tested everything yet but so far it's as you said: it just works without having to change much.

newtonick commented:
tACK
